### PR TITLE
haskell: Add uhc compiler

### PR DESCRIPTION
**Pull request by phile314**
Adds the UHC haskell compiler.

Based upon the previously removed uhc package (see issue #3565 ), tested on Fedora 21 amd64.

### CODE REVIEW

File: pkgs/development/compilers/uhc/default.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{ stdenv, coreutils, fetchgit, m4, libtool, clang, ghcWithPackages,
+  shuffle,
+  binary, hashable, mtl, network, uhc-util, uulib
+}:
+
+let w = ghcWithPackages ( self: [binary hashable mtl network uhc-util uulib] );
+    wrappedGhc = w.override { ignoreCollisions = true; };
```
Comment (peti) [inline pkgs/development/compilers/uhc/default.nix]: Why is the `ignoreCollisions` override necessary?

File: pkgs/development/compilers/uhc/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    sed -i "s|--make|--make -package-db=$p|g" src/ehc/files2.mk
+    sed -i "s|--make|--make -package-db=$p|g" src/gen/files.mk
+  '';
+
+  inherit clang;
+
+  meta = with stdenv.lib; {
+    homepage = "http://www.cs.uu.nl/wiki/UHC";
+    description = "Utrecht Haskell Compiler";
+    maintainers = [ maintainers.phausmann ];
```
Comment (peti) [inline pkgs/development/compilers/uhc/default.nix]: Please don't assign maintainership of this package to someone else.

File: pkgs/development/compilers/uhc/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    platforms = stdenv.lib.platforms.linux;
```
Comment (peti) [inline pkgs/development/compilers/uhc/default.nix]: Why are builds restricted to Linux? Can Darwin not compile this package?

File: pkgs/development/compilers/uhc/default.nix (hunk 2 above)
Comment (peti) [inline pkgs/development/compilers/uhc/default.nix]: Duh, sorry. I misread the maintainer line. Please disregard my earlier comment.

**Comment by phile314** [comment]
I made the changes. This is my first nix package, so sorry for any silly mistakes...

I use platforms.unix now. I know that UHC works on linux and Mac OS X, probably also on the *bsd's, so I guess that should be okay?

File: pkgs/development/compilers/uhc/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{ stdenv, coreutils, fetchgit, m4, libtool, clang, ghcWithPackages,
+  shuffle,
+  binary, hashable, mtl, network, uhc-util, uulib
+}:
+
+let wrappedGhc = ghcWithPackages ( self: [binary hashable mtl network uhc-util uulib] );
```
Comment (peti) [inline pkgs/development/compilers/uhc/default.nix]: Why is it necessary to pass `binary` here? That library is a core package, i.e. it comes with `ghc` anyway.

**Comment by phile314** [comment]
I removed the explicit dependency on binary.

**Pull request merged**